### DISCUSSION TITLE
Operator assumes security plugin to be enabled always

### DISCUSSION THREAD

**Discussion by gaurav-05** [Opster/opensearch-k8s-operator#396] [General]
Hi.. Currently operator assumes that security plugin is always enabled in the image being used, but we can have a case when we do not want security plugin to be enabled, is there a way right now we can disable the TLS logic from being called and use HTTP instead.

**Comment by swoehrl-mw** [comment]
Hi @gaurav-05. There is currently no way to disable TLS via the operator and, as you've already deduced, the operator assumes in several places that TLS and the securityplugin are always enabled. So even if you build a custom image without the securityplugin it would not work.

To be honest I don't see this being supported in the future. One of the main goals of the operator is to provide (nearly) production-ready clusters, and TLS and authentication is a must-have for that.

**Reply by gaurav-05** [reply]
Thanks @swoehrl-mw for your reply. Yes I see that for making it run with my custom image(without security plugin) I had to do some tweaks to the code.

But I am not sure that having security plugin would be must for production ready clusters, like we are having istio in our cluster and rely only on that for cluster security.

Also I see we can not have dynamic check as it assumes https at places by default again, so another way could be having some flag which is true by default and set to false when you don't have plugin. Or you may suggest some way - I can submit the PR if you think it can make in.

**Reply by swoehrl-mw** [reply]
> But I am not sure that having security plugin would be must for production ready clusters, like we are having istio in our cluster and rely only on that for cluster security.

The security plugin also provides authentication (both for users via the REST API and for node2node communication). These are things that IMO you cannot substitute with something like istio and are must-haves for real clusters.

What I could see as a posibility is providing a flag to disable TLS on the REST API layer but keeping the security plugin itself and TLS for the transport layer active. But even there I feel that having it active all the time and accepting that in some situations (like in your case with istio) there is double encryption (which nowadays should not be a big performance deal) is best for the majority of users.

**Reply by gaurav-05** [reply]
Adding more problem to this we also use keycloak for authentication so we do not want to use default authentication for which reason we do not use security plugin...thoughts ?

**Reply by gaurav-05** [reply]
@swoehrl-mw Any thoughts on above also can you suggest if how we can disable TLS on REST API only.

**Reply by swoehrl-mw** [reply]
Hi @gaurav-05. As I already stated, disabling TLS is currently not possible with the operator and I don't know of any workarounds.
If you want to use Keycloak for authentication, I don't see why you would need to disable the security plugin? The normal way to integrate keycloak would be via Oauth/OIDC and have only some technical users in the internal_users list. Both is possible with the operator.

**Reply by gaurav-05** [reply]
Sure and would you allow any PR that provides option to disable TLS.

**Reply by swoehrl-mw** [reply]
I can accept an option that allows disabling TLS on the REST layer (with the default being TLS enabled and a clear warning not to do it unless other measures are in place). Completely disabling TLS and/or the security plugin is something we IMO should not support.

**Reply by roelandvanbatenburg** [reply]
We are not exposing the cluster in anyway and run OS with both `DISABLE_INSTALL_DEMO_CONFIG` and `DISABLE_SECURITY_PLUGIN`. The first thing we are experiencing is that the readinessProbe only communicates via https and thus none of the pods are marked as ready.

**Reply by swoehrl-mw** [reply]
Hi @roelandvanbatenburg. As already discussed above the operator does not support disabling TLS or the entire security plugin. As such the behaviour you describe is expected.
